### ISSUE TITLE
CastPlayer.getDuration() returns -1000 for live streams

### ISSUE THREAD

**Issue by Stonos** [google/ExoPlayer#7983]
### [REQUIRED] Issue description
`CastPlayer.getDuration()` always seems to return `-1000` when casting a live stream. This duration eventually becomes  the `TimeBar`'s duration, which causes the scrubber to disappear, leaving you unable to timeshift the live stream.

I have tried setting the `StreamType` to `MediaInfo.STREAM_TYPE_LIVE`, but it didn't help.

A possible workaround seems to be:
```kotlin
override fun getDuration(): Long {
    val remoteMediaClient = remoteMediaClient ?: return super.getDuration()
    val seekableDuration = remoteMediaClient.approximateLiveSeekableRangeEnd - remoteMediaClient.approximateLiveSeekableRangeStart

    return if (seekableDuration > 0) {
        seekableDuration
    } else {
        super.getDuration() // seekableDuration is 0 for VODs
    }
}
```

However, this doesn't solve the `TimeBar` issue, since its duration is based on the `Timeline`'s duration.

### [REQUIRED] Reproduction steps
1. Cast a live stream
2. Call `CastPlayer.getDuration()`
3. The result of the above above function call is `-1000`

### [REQUIRED] Link to test content
https://live.unified-streaming.com/scte35/scte35.isml/.mpd?dvr_window_length=3600

### [REQUIRED] A full bug report captured from the device
N/A

### [REQUIRED] Version of ExoPlayer being used
2.12.0

### [REQUIRED] Device(s) and version(s) of Android being used
* LG H815 Android 9
* Android 8 TV with built-in Chromecast

**Comment by AquilesCanta** [comment]
Thanks for reporting and apologies for the delay. Reported the bug to the cast team [Internal ref: b/171657375] and I'll probably add a workaround in our codebase as well.

Aside, is the stream playing correctly for you?

**Comment by Stonos** [comment]
Yes, other than that the stream is playing fine.

One thing that is not working correctly however is timeshift. Even with the workaround applied, timeshift was not working correctly (the time was moving too fast and seeking would not seek to the correct position). I had to override the following methods to make timeshift work properly for live streams:

```kotlin
override fun seekTo(windowIndex: Int, positionMs: Long) {
    val approximateLiveSeekableRangeStart = remoteMediaClient?.approximateLiveSeekableRangeStart
            ?: 0
    super.seekTo(windowIndex, positionMs + approximateLiveSeekableRangeStart)
}

override fun getCurrentPosition(): Long {
    return if (isCurrentWindowLive) { // fix current position on timeshift
        val approximateLiveSeekableRangeStart = remoteMediaClient?.approximateLiveSeekableRangeStart
                ?: 0
        super.getCurrentPosition() - approximateLiveSeekableRangeStart
    } else {
        super.getCurrentPosition()
    }
}
```

(in order to do that, I modified the original `CastPlayer` class to add a getter for `remoteMediaClient` and made it non-final so that I can extend it)

**Comment by AquilesCanta** [comment]
I think we should incorporate these workarounds (or equivalent changes) to the default implementation. I'll see what I can do about it. My issue right now is that the stream does not play for me.

Which application ID are you depending on? 

I'm using the dev branch's "A12D4273" app id. If you are not using this one, would you mind trying it out?

**Comment by Stonos** [comment]
Just tested app ID `A12D4273` using the Cast demo app on the development branch, and it worked almost fine (the stream started from the beginning instead of the live edge) on a Chromecast 2 and on an Android TV built-in Chromecast.

These are the changes I made to the demo app:
```patch
Index: demos/cast/src/main/java/com/google/android/exoplayer2/castdemo/DemoUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- demos/cast/src/main/java/com/google/android/exoplayer2/castdemo/DemoUtil.java	(revision 5b1514e933f1bd7415737af3c2323f47f08f0b64)
+++ demos/cast/src/main/java/com/google/android/exoplayer2/castdemo/DemoUtil.java	(date 1603711334733)
@@ -41,6 +41,12 @@
     // Clear content.
     samples.add(
         new MediaItem.Builder()
+            .setUri("https://live.unified-streaming.com/scte35/scte35.isml/.mpd?dvr_window_length=3600")
+            .setMediaMetadata(new MediaMetadata.Builder().setTitle("Live stream").build())
+            .setMimeType(MIME_TYPE_DASH)
+            .build());
+    samples.add(
+        new MediaItem.Builder()
             .setUri("https://storage.googleapis.com/wvmedia/clear/h264/tears/tears.mpd")
             .setMediaMetadata(new MediaMetadata.Builder().setTitle("Clear DASH: Tears").build())
             .setMimeType(MIME_TYPE_DASH)
Index: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/DefaultMediaItemConverter.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/DefaultMediaItemConverter.java	(revision 5b1514e933f1bd7415737af3c2323f47f08f0b64)
+++ extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/DefaultMediaItemConverter.java	(date 1603711561578)
@@ -62,7 +62,7 @@
     }
     MediaInfo mediaInfo =
         new MediaInfo.Builder(item.playbackProperties.uri.toString())
-            .setStreamType(MediaInfo.STREAM_TYPE_BUFFERED)
+            .setStreamType(MediaInfo.STREAM_TYPE_LIVE)
             .setContentType(item.playbackProperties.mimeType)
             .setMetadata(metadata)
             .setCustomData(getCustomData(item))
```

I'm not really sure why the stream is not working for you 😕

**Comment by AquilesCanta** [comment]
Thanks for checking. For context, the provided app id is the same as the default receiver except it provides us with usage metrics, so it would be great if you could use it for your apps. Usage metrics helps us allocate resources to fixing Cast issues.

**Comment by Stonos** [comment]
I'd like to, but unfortunately I cannot use it in my app because I'm already using another customized custom receiver.

**Comment by condesales** [comment]
I'm currently observing the same issue on 2.12.1

I have a custom controls view that extends from the default `PlayerControlView`.
The controls view then receives a reference for the player via
```kotlin
controls.player = player
```
Where the `player` instance can vary between `SimpleExoPlayer` and `CastPlayer`
For SimpleExoplayer the controls work fine for either VOD or LIVE but when using the CastPlayer it only works as expected for VOD

Once I plugin in a live stream, the `DefaultTimeBar` view does not show duration or position.
CastPlayer.duration always retuns -1000

I've currently implemented a workaround using `remoteMediaClient.addProgressListener` to receive updates on progress and force update the DefaultTimeBar values using
```kotlin
val duration = remoteMediaClient.approximateLiveSeekableRangeEnd - remoteMediaClient.approximateLiveSeekableRangeStart
val progress = remoteMediaClient.approximateStreamPosition
timebar.setDuration(duration)
timebar.setPosition(progress)
```
But I cannot say I'm very proud of that solution

There hasn't been a lot of updates on this issue, so let me know if there's any further information you need to help to solve it.

**Comment by condesales** [comment]
I also believe #7166 is related to this one

**Comment by AquilesCanta** [comment]
@condesales The fix for this was introduced in 2.12.2, so you'll have to update to pick it up. Closing.

See:
https://github.com/google/ExoPlayer/blob/r2.12.2/extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastUtils.java.
